### PR TITLE
Cspu hybrid 20190626

### PR DESCRIPTION
**Pull request by cfmcginn**
Describe the Pull-Request: Hybrid CS-PU jet reconstruction w/ flow correction and better background eta control

Does the PR pass the test script located at
HeavyIonsAnalysis/JetAnalysis/test/runtest.sh
[x] Yes
[] No

Please make sure to mention (using the @ syntax) anyone who might be interested in this PR.
@bi-ran, @mandrenguyen

### CODE REVIEW

File: HeavyIonsAnalysis/JetAnalysis/test/runForestAOD_PbPb_MIX_75X.py [hunk 1]
```diff
@@ -101,7 +101,9 @@
 
 
 #PU minimal tower cut reco sequence
-process.load('HeavyIonsAnalysis.JetAnalysis.FullJetSequence_puLimitedPbPb')
+#process.load('HeavyIonsAnalysis.JetAnalysis.FullJetSequence_puLimitedPbPb')
+# EDIT HERE FOR CSPU HYBRID
+process.load('HeavyIonsAnalysis.JetAnalysis.FullJetSequence_appendCSPUHybrid')
```
Comment (bi-ran) [inline HeavyIonsAnalysis/JetAnalysis/test/runForestAOD_PbPb_MIX_75X.py:106]: @cfmcginn this file seems to be missing from this pr?

**Comment by cfmcginn** [comment]
Two latest commits fix this issue and add data version, which I realized was missing. I did a fresh pull of this and i believe now everything is correct. runtest.sh is passed

**Comment by bi-ran** [comment]
@cfmcginn thanks!

**Comment by mandrenguyen** [comment]
@cfmcginn @bi-ran What is the timescale for forward porting this into 10_3_X?

**Comment by cfmcginn** [comment]
Assuming no complications, EOD today

**Comment by cfmcginn** [comment]
https://github.com/cfmcginn/cmssw/tree/CSPUHybrid_20190715

^103X branch there - in modding code for 103X I introduced some sort of memory leak. Debugging now, PR when solved

**Comment by FHead** [comment]
Hi @cfmcginn can you describe the symptom for memory leak?  If it will still take some time, do you want someone to take look and look into it?

**Comment by cfmcginn** [comment]
Hi @FHead, bug is in PURhoProducer, still trying to isolate what part of module. Feel free to take a look at the .cc file

**Comment by FHead** [comment]
Thanks, can you point me to an example setup + config file to reproduce the problem?

**Comment by cfmcginn** [comment]
https://github.com/cfmcginn/cmssw/tree/CSPUHybrid_20190715 This branch, run w/ pponAA MIX python should do

**Comment by cfmcginn** [comment]
Bug is here, ietamin and ietamax are randomly not set properly
https://github.com/cfmcginn/cmssw/blob/CSPUHybrid_20190715/RecoHI/HiJetAlgos/plugins/HiPuRhoProducer.cc#L309

**Comment by FHead** [comment]
Nice!  We can do the PR then?

**Comment by cfmcginn** [comment]
This is just bug location - do not know why they are randomly not set yet. Looking at it now

**Comment by FHead** [comment]
Is this an issue?  When will the pointers get freed?
https://github.com/cfmcginn/cmssw/blob/CSPUHybrid_20190715/RecoHI/HiJetAlgos/plugins/HiPuRhoProducer.cc#L169-L170

**Comment by cfmcginn** [comment]
Very possible I missed the free - I can go ahead and fix - Someone should backport to 758p3 forest along with the =NULL fix

**Comment by cfmcginn** [comment]
Branch is updated w/ delete statements.
https://github.com/cfmcginn/cmssw/blob/CSPUHybrid_20190715/RecoHI/HiJetAlgos/plugins/HiPuRhoProducer.cc#L184
 I cannot test unfortunately so someone should pull and check nothing breaks. Also this is based on our own iterative pedestal code so someone should check there is no leak there.

**Comment by bi-ran** [comment]
those are typedef'd boost::shared_ptr, so delete does not work on them (cannot compile). could you please revert the last commit? the memory of shared_ptrs will be managed automatically, so there's no issue with leaks there.

**Comment by FHead** [comment]
The problem is that the pointer is saved in the class member, so the object doesn't get destroyed after each event (which is the trigger for memory release).  Does the delete of the previous one happen at the assignment of the next event?

**Comment by cfmcginn** [comment]
I guess before I revert:

https://github.com/cms-sw/cmssw/blob/master/RecoHI/HiJetAlgos/src/MultipleAlgoIterator.cc#L25

It is based on here (our PU algo), and as best I can tell there is no additional handling of the clustersequenceptr

**Comment by FHead** [comment]
Yes indeed it's destroyed at the next assignment - let's revert

**Comment by cfmcginn** [comment]
Reversion is done.

**Pull request merged**